### PR TITLE
Refactor SavedVariable

### PR DESCRIPTION
**Pull request by Varal7**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59838 Remove dead code in SavedVariable
* #59837 Avoid un-necessary unwrapping of Tensor in SavedVariable
* **#59836 Refactor SavedVariable**

Summary: Preparing for #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069159](https://our.internmc.facebook.com/intern/diff/D29069159)

### CODE REVIEW

File: torch/csrc/autograd/saved_variable.cpp [hunk 1]
```diff
@@ -37,28 +37,34 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
       "you can make a clone to get a normal tensor and use it in autograd.")
 
     was_default_constructed_ = false;
+    is_inplace_view_ = is_inplace_view;
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:40]: nit: following the discussion we had, could you rename this in `is_inplace_on_view`? to reflect that we're not talking about inplace view operations but inplace operations performed on a view?

File: torch/csrc/autograd/saved_variable.cpp [hunk 2]
```diff
@@ -67,19 +73,15 @@ SavedVariable::SavedVariable(const c10::optional<Variable>& variable, bool is_ou
...
 Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
   if (!data_.defined()) {
-    if (!was_default_constructed_) {
-      throw std::runtime_error(ERR_BACKWARD_TWICE);
-    }
+    TORCH_CHECK(was_default_constructed_, ERR_BACKWARD_TWICE);
     return Variable();
   }
 
+  // We want grad_fn here to provide the most hlepful debug message to the user
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:80]: nit: helpful

**Review by albanD — APPROVED** [review]
LGTM

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069159).

**Pull request closed without merging**